docs: initial content for Using packages (#16)
NativeScript/docs: content/guide/development-workflow/using-packages.md

---

> title: Using packages in NativeScript

Reviewer: ```suggestion
title: Using packages (aka, plugins) in NativeScript
```

---

### Plugins

> NativeScript plugins are npm packages with some added native functionality. Therefore, finding, installing, and removing NativeScript plugins works a lot like working with npm packages you might use in your Node.js or front-end web development.

Reviewer 2: ```suggestion
NativeScript plugins are npm packages that often include additional native functionality. Consequently, discovering, installing, and removing NativeScript plugins resembles the process of working with npm packages used in Node.js or other front-end web development.
```

#### Finding plugins

The NativeScript team maintains a list of plugins [here](https://docs.nativescript.org/plugins/index.html).

> You can also look for plugins at the [NativeScript marketplace](https://market.nativescript.org/) or on the [npm’s site](https://www.npmjs.com/)

Reviewer 2: ```suggestion
Plugins can also be found at the [NativeScript marketplace](https://market.nativescript.org/) or on the [npm’s site](https://www.npmjs.com/)
```

> Not finding the right plugin? You can try asking on our [Discord Community](https://nativescript.org/discord). Or perhaps [build a new plugin](/https://docs.nativescript.org/plugins/index.html#developing-your-own-plugins).

Reviewer 2: ```suggestion
If the desired plugin is not found, the option is available to get in touch with the Nativescript discord group at [Discord Community](https://nativescript.org/discord), or alternatively at [build a new plugin](https://docs.nativescript.org/plugins/index.html#developing-your-own-plugins).
```

Reviewer 2: TODO can be removed

<!-- TODO: fix links -->

> Also, make sure to look through the [NativeScript core modules](https://docs.nativescript.org/understanding-packages.html#nativescript-core), which ship as a dependency of every NativeScript app. There’s a chance that the functionality you need is built in. 

Reviewer 2: ```suggestion
You can also explore [@nativescript/core](https://docs.nativescript.org/understanding-packages.html#nativescript-core), which is included as a dependency in every NativeScript app. It is possible that the required functionality is already available.
```

#### Installing Plugins

> Once you’ve found the plugin you need, install the plugin into your app using the `ns plugin add` command.

Reviewer 2: ```suggestion
Once the needed plugin is found, it can be installed in a NativeScript app using the `ns plugin add` command which behaves just like `npm install` while also modifying any existing platform projects (linking, etc.).
```

```cli
ns plugin add <plugin-name>
```

For example, the following command installs the [NativeScript camera plugin](https://docs.nativescript.org/plugins/camera.html).

```cli
ns plugin add @nativescript/camera
```

> Instead of using `plugin add`, you can use your package manager as well (npm, yarn, pnpm...):

Reviewer 2: ```suggestion
Instead of using plugin add, package managers like npm, yarn, or pnpm can also be used:
```

```cli
npm install --save @nativescript/camera
```

> The installation of a NativeScript plugin mimics the installation of an npm package. The NativeScript CLI downloads the plugin from npm and adds the plugin to the `node_modules` folder in the root of your project. During this process, the NativeScript CLI adds the plugin to your project’s root `package.json` file and also resolves the plugin’s dependencies (if any).

Reviewer 2: ```suggestion
The installation of a NativeScript plugin is similar to installing a npm package.  The NativeScript CLI downloads the plugin from npm and adds it to the node_modules folder in the project's root. This process also includes updating the root package.json file with the plugin and resolving its dependencies (if any).
```

#### Installing Plugins as Developer Dependencies

> As shown above the command `ns plugin add @nativescript/camera` is actually doing `npm i @nativescript/camera --save` behind the scenes. If you need to install a **developer dependency** in your project (e.g., like **@nativescript/types** or **@nativescript/webpack**) then you will need to explicitly save it as a **devDependency**. To achieve that, use the `npm install` command with `--save-dev` flag. For example:

Reviewer 2: ```suggestion
As demonstrated above, the command `ns plugin add @nativescript/camera` is equivalent to `npm i @nativescript/camera --save` internally.  However, when installing a developer dependency, such as @nativescript/types or @nativescript/webpack, you need to explicitly save it as a devDependency.  To do so, use the npm install command with the --save-dev flag. For instance:
```

```cli
npm i @nativescript/types --save-dev
```

::: tip Note

> The difference between dependencies and developer dependencies is that **dependencies** are required to run, while **devDependencies** are needed only during development. Example for dependency is the **@nativescript/camera** plugin which is required at runtime so you could use the hardware camera. On the other hand, the **@nativescript/types** is a developer dependency required only for intelliSense during the development process. The `devDependencies` should not be installed as `dependencies` to avoid large output build files (large application size). Example `package.json` file using both `dependencies` and `devDependencies` can be found [here](https://github.com/NativeScript/nativescript-sdk-examples-js/blob/master/package.json#L31-L44).

Reviewer 2: ```suggestion
Regarding package handling, **dependencies** are necessary for the application to run, while **devDependencies** are only needed during the development phase. For instance, the **@nativescript/camera** plugin is a dependency required at runtime to utilize the hardware camera.  Conversely, **@nativescript/types** is a developer dependency used solely for IntelliSense during the development process. It is important to avoid installing `devDependencies` as `dependencies` to prevent large output build files and keep the application size manageable. You can find an example of a `package.json` file using both `dependencies` and `devDependencies` [here](https://github.com/NativeScript/nativescript-sdk-examples-js/blob/master/package.json#L31-L44).
```

:::

#### Importing and Using Plugins

> Once the plugin you need is installed, you can start using it in your project. Note that each plugin might have its configuration that needs to be satisfied so always check carefully the plugin's documentation and the README file. The below code snippet demonstrated the basic usage of **@nativescript/camera** plugin.

Reviewer 2: ```suggestion
Once the required plugin is installed, it can be utilized in the project. It's crucial to carefully review the plugin's documentation and README file as each plugin may have specific configuration requirements. The following code snippet demonstrates the basic usage of the [@nativescript/camera](https://docs.nativescript.org/plugins/camera.html) plugin:
```

```javascript
import { requestPermissions } from '@nativescript/camera'
requestPermissions()
```

```typescript
import { requestPermissions } from '@nativescript/camera'
requestPermissions()
```

#### Removing Plugins

> To remove a NativeScript plugin from your project, run the following command from your command line.

Reviewer 2: ```suggestion
To remove a NativeScript plugin, you can use the following command.
```

```cli
ns plugin remove <plugin-name>
```

For example, the following command removes the NativeScript camera plugin.

```cli
ns plugin remove @nativescript/camera
```

As with installation, the removal of a NativeScript plugin mimics the removal of an npm package.

> The NativeScript CLI removes any plugin files from your app’s `node_modules` folder in the root of your project. The CLI also removes any of the plugin’s dependencies and also removes the plugin from your project’s root `package.json` file.

Reviewer 2: ```suggestion
The NativeScript CLI removes plugin files from the app's node_modules folder in the project's root. It also removes the plugin's dependencies and removes the plugin from the project's root package.json file.
```

### Package Managers

> A package manager is a piece of software that lets you manage the external code, written by you or someone else, that your project needs to work correctly. By default, NativeScript CLI uses Node Package Manager (`npm`) for managing the dependencies of the application. When new application is created, CLI automatically calls `npm install` to install all of its dependencies.

Reviewer 2: ```suggestion
A package manager is a piece of software that helps manage the external code needed for your project to work correctly.  By default, NativeScript CLI uses Node Package Manager (npm) for managing the application's dependencies.  When a new application is created, CLI automatically calls npm install to install all of the required dependencies.
```

> A package manager is a piece of software that lets you manage the external code, written by you or someone else, that your project needs to work correctly. By default, NativeScript CLI uses Node Package Manager (`npm`) for managing the dependencies of the application. When new application is created, CLI automatically calls `npm install` to install all of its dependencies.

Reviewer 2: ```suggestion
A package manager is software that enables management of external code necessary for a project to function properly.  The NativeScript CLI employs Node Package Manager (npm) as its default package manager.  Upon creating a new app, the CLI automatically executes `npm install` to install all required dependencies.
```
